sdk/metric/export/data.go

```diff
@@ -20,6 +20,131 @@
+	Data Aggregation
+}
+
+// Aggregation is the store of data reported by an Instrument.
+// It will be one of: Gauge, Sum Histogram.
+type Aggregation interface {
+	privateAggregation()
+}
+
+// Gauge represents a measurement of the current value of an instrument.
+type Gauge struct {
+	// DataPoints reprents individual aggregated measurements with unique Attributes.
+	DataPoints []DataPoint
+}
+
+func (Gauge) privateAggregation() {}
+
+// Sum represents the sum of all measurements of values from an instrument.
+type Sum struct {
+	// DataPoints reprents individual aggregated measurements with unique Attributes.
+	DataPoints []DataPoint
+	// Temporality describes if the aggregation is reported as the change from the
+	// last report time, or the cumulative changes since a fixed start time.
+	Temporality Temporality
+	// IsMonotonic represents if this aggregation only increases or decreases.
+	IsMonotonic bool
+}
+
+func (Sum) privateAggregation() {}
+
+// DataPoint is a single data point in a timeseries.
+type DataPoint struct {
+	// Attributes is the set of key value pairs that uniquely identify the timeseries.
+	Attributes []attribute.KeyValue
+	// StartTime is when the timeseries was started. (optional)
+	StartTime time.Time
+	// Time is the time when the timeseries was recorded. (optional)
+	Time time.Time
+	// Value is the value of this data point.
+	Value Value
+}
+
+// Value is a int64 or float64. All Values created by the sdk will be either
+// Int64 or Float64.
+type Value interface {
+	privateValue()
+}
+
+// Int64 is a container for an int64 value.
+type Int64 int64
+
+func (Int64) privateValue() {}
+
+// Float64 is a container for a float64 value.
+type Float64 float64
+
+func (Float64) privateValue() {}
+
+// Histogram represents the histogram of all measurements of values from an instrument.
+type Histogram struct {
+	// DataPoints reprents individual aggregated measurements with unique Attributes.
+	DataPoints []HistogramDataPoint
+	// Temporality describes if the aggregation is reported as the change from the
+	// last report time, or the cumulative changes since a fixed start time.
+	Temporality Temporality
+}
+
+func (Histogram) privateAggregation() {}
+
+// HistogramDataPoint is a single histogram data point in a timeseries.
+type HistogramDataPoint struct {
+	// Attributes is the set of key value pairs that uniquely identify the timeseries.
+	Attributes []attribute.KeyValue
+	// StartTime is when the timeseries was started. (optional)
+	StartTime time.Time
+	// Time is the time when the timeseries was recorded. (optional)
+	Time time.Time
+
+	// Count is the number of updates this histogram has been calculated with.
+	Count uint64
+	// Bounds are the upper bounds of the buckets of the histogram. Because the
+	// last boundary is +infinity this one is implied.
+	Bounds []float64
+	// BucketCounts is the count of each of the buckets.
+	BucketCounts []uint64
+
+	// Min is the minimum value recorded. (optional)
+	Min float64
```

Comment:
do we need optional attributes (start time/time/min/max/sum) which can be "unset" to be pointers so we can differentiate between set to 0 and unset?

Reply:
I would actually think that we don't need to mark these as optional.  We just need to have the different aggregations always populate them.  They are only optional for the otlp wireformat.

Would you rather remove the optional verbiage, or move these to pointers?

Reply:
Min/Max setting is configurable, IIRC, so I think those need to be optional, and pointers.  I think start time/end time/sum should always be populated so those can probably have optional removed?

Reply:
For the SDK derivative in otel-launcher-go, I've used `NaN` values to indicate when Min and Max are not set. Generally, I would expect `Min` and `Max` values of the data point to be valid whenever `Count != 0`. It's a separate question whether the OTLP exporter encodes Min/Max or not. In the otel-launcher-go repo we have a MinMaxSumCount aggregator that encodes Min/Max fields only when using Delta temporality; admittedly, this is not exactly specified but the datamodel points out why Min/Max are not considered all that useful for Cumulative temporality.